pkg/queue/rabbitmq/internal/connection/consumer.go

```diff
@@ -42,144 +45,198 @@ var th2RabbitmqMessageProcessDurationSeconds = promauto.NewHistogramVec(
-		conn:     conn,
-		channels: make(map[string]*amqp.Channel),
-		Logger:   logger,
-	}, nil
+	return consumer, nil
 }
 
-func (cns *Consumer) Close() error {
-	return cns.conn.Close()
+func (cns *Consumer) Consume(queueName string, th2Pin string, th2Type string, handler func(delivery amqp.Delivery) error) error {
+	return cns.consume(
+		queueName, th2Pin, th2Type, cns.consumeWithAutoAck, "consume",
+		func(delivery amqp.Delivery, timer *prometheus.Timer) error {
+			defer timer.ObserveDuration()
+			return handler(delivery)
+		},
+	)
 }
 
-func (cns *Consumer) Consume(queueName string, th2Pin string, th2Type string, handler func(delivery amqp.Delivery) error) error {
-	ch, err := cns.conn.Channel()
+func (cns *Consumer) ConsumeWithManualAck(queueName string, th2Pin string, th2Type string, handler func(msgDelivery amqp.Delivery, timer *prometheus.Timer) error) error {
+	return cns.consume(
+		queueName, th2Pin, th2Type, cns.consumeWithManualAck, "consumeWithManualAck",
+		func(delivery amqp.Delivery, timer *prometheus.Timer) error {
+			return handler(delivery, timer)
+		},
+	)
+}
+
+func (cns *Consumer) consume(queueName string, th2Pin string, th2Type string,
+	subscriptionProducer func(queueName string, methodName string) (*amqp.Channel, <-chan amqp.Delivery, error),
+	methodName string, handler func(delivery amqp.Delivery, timer *prometheus.Timer) error) error {
+	ch, msgs, err := subscriptionProducer(queueName, methodName)
 	if err != nil {
-		cns.Logger.Error().
-			Err(err).
-			Str("queue", queueName).
-			Msg("cannot open channel")
 		return err
 	}
-	cns.channels[queueName] = ch
-
-	msgs, consErr := ch.Consume(
-		queueName, // queue
-		// TODO: we need to provide a name that will help to identify the component
-		"",    // consumer
-		true,  // auto-ack
-		false, // exclusive
-		false, // no-local
-		false, // no-wait
-		nil,   // args
-	)
-	if consErr != nil {
-		cns.Logger.Error().
-			Err(err).
-			Str("method", "consume").
-			Str("queue", queueName).
-			Msg("Consuming error")
-		return consErr
-	}
 
 	go func() {
 		cns.Logger.Debug().
-			Str("method", "consume").
+			Str("method", methodName).
 			Str("queue", queueName).
 			Msg("start handling messages")
-		for d := range msgs {
+		running := true
+		handleDelivery := func(d amqp.Delivery) {
 			timer := prometheus.NewTimer(th2RabbitmqMessageProcessDurationSeconds.WithLabelValues(th2Pin, th2Type, queueName))
 			cns.Logger.Trace().
 				Str("exchange", d.Exchange).
 				Str("routing", d.RoutingKey).
 				Int("bodySize", len(d.Body)).
 				Msg("receive delivery")
-			if err := handler(d); err != nil {
+			if err := handler(d, timer); err != nil {
 				cns.Logger.Error().
 					Err(err).
 					Str("exchange", d.Exchange).
 					Str("routing", d.RoutingKey).
 					Int("bodySize", len(d.Body)).
 					Msg("Cannot handle delivery")
 			}
-			timer.ObserveDuration()
 			th2RabbitmqMessageSizeSubscribeBytes.WithLabelValues(th2Pin, th2Type, queueName).Add(float64(len(d.Body)))
```

Comment:
Could we hold values returned from WithLabelValues method?

Reply:
Good idea. According to the documentation, we can. Will store observer objects in local variables